### PR TITLE
Scale with keda

### PR DESCRIPTION
**Pull request by kwiatekus**
**Description**

Changes proposed in this pull request:

- Example illustrating how to use functions, eventing and keda to build an application composed of event driven, scallable functions 

**Related issue(s)**
https://github.com/kyma-project/kyma/issues/11225

### CODE REVIEW

File: scale-to-zero-with-keda/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+## Overview
+This example demostrates a very useful setup utilising
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:2]:
```suggestion
This example demonstrates a handy setup utilizing: 
```
Can you provide more info about the setup? What does it set up exactly?

File: scale-to-zero-with-keda/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
+## Overview
+This example demostrates a very useful setup utilising 
+ - Functions to deploy workloads directly from git repository,
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:3]:
```suggestion
 - Functions to deploy workloads directly from a Git repository,
```

File: scale-to-zero-with-keda/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ - In-cluster Eventing to enable event-driven communication, 
+ - [Keda](https://keda.sh/) to drive Function scaling,
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:5]:
```suggestion
 - [Keda](https://keda.sh/) to drive the Function's scaling,
```

File: scale-to-zero-with-keda/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ - Prometheus and Istio to deliver metrics essential for scaling decisions.
+
+It realises the following scenario:
+
+![scenario](./assets/scaling-scenario.png "Scenario")
+
+The proxy Function receives incomming http trafic and with every request it publishes the payload as incluster event to a particular topic.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:12]:
```suggestion
The proxy Function receives the incoming HTTP traffic, and with every request, it publishes the payload as an in-cluster event to a particular topic.
```

File: scale-to-zero-with-keda/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+The second Function ( the actual worker ) is subscribed to the topic and processes incomming  messages. But until there are no messages published for it's subscribed topic it remains scaled to zero - there are no actual worker Pods living in the runitme.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:14]:
```suggestion
The second Function (the actual worker) is subscribed to the topic and processes the incoming messages. Until there are no messages published for its subscribed topic, it remains scaled to zero - there are no actual worker Pods living in the runtime.
```

File: scale-to-zero-with-keda/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+Keda is used to scale the worker Function. [KEDA Prometheus scaler](https://keda.sh/docs/2.8/scalers/prometheus/) is used to measure the load targeted for the worker's Function and scale it accordingly ( from 0 to 5 replicas )
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:16]:
```suggestion
Keda is used to scale the worker Function. [KEDA Prometheus scaler](https://keda.sh/docs/2.8/scalers/prometheus/) is used to measure the load targeted for the worker Function and scale it accordingly (from 0 to 5 replicas).
```

File: scale-to-zero-with-keda/README.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+
+## Prerequisites
+
+- Kyma as the target deployment environment.
+
+## Installation
+
+Run the following against your kyma runtime to install KEDA:
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:25]:
```suggestion
Run the following against your Kyma runtime to install KEDA:
```
What do you mean by `run against`?

File: scale-to-zero-with-keda/README.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+```bash
+kubectl apply -f https://github.com/kedacore/keda/releases/download/v2.8.0/keda-2.8.0.yaml
+```
+
+Make sure istio sidecar injection is enabled in the target namesapce:
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:31]:
```suggestion
Make sure istio sidecar injection is enabled in the target Namesapce:
```

File: scale-to-zero-with-keda/README.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+Apply the example resources from `./k8s-resources` directory:
+```bash
+kubectl apply -f ./k8s-resources`
+```
+
+## Test the application
+
+At first the worker Function is scaled down.
+When listing HPA for the Function you will see that current replica count is at zero.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:45]:
```suggestion
When listing HPA for the Function, you will see that the current replica count is zero.
```

File: scale-to-zero-with-keda/README.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+ ```bash
+kubectl get hpa
+NAME                               REFERENCE                     TARGETS             MINPODS   MAXPODS   REPLICAS   AGE
+keda-hpa-worker-fn-scaled-object   Function/scalable-worker-fn   <unknown>/2 (avg)   1         5         0          27h
+
+ ```
+ Also, when listing Pods by Function name label you will see only the build job's Pod. No runtime Pod is up.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:52]:
```suggestion
 Also, when listing Pods by Function name label, you will see only the build job's Pod. No runtime Pod is up.
```

File: scale-to-zero-with-keda/README.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+ ```bash
+kubectl get pods -l serverless.kyma-project.io/function-name=scalable-worker-fn -w
+NAME                                   READY   STATUS      RESTARTS   AGE
+scalable-worker-fn-build-7s4rf-wjhvt   0/1     Completed   0          2m16s
+ ```
+
+Once you generate a load ( even a single request ) the non-zero request rate targeting the worker Function will trigger scaling up of the worker Function's runtime Pods.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:59]:
```suggestion
Once you generate a load (even a single request), the non-zero request rate targeting the worker Function triggers scaling up of the worker Function's runtime Pods.
```

File: scale-to-zero-with-keda/README.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+ Call the http proxy Function once
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:61]:
```suggestion
 Call the HTTP proxy Function once:
```

File: scale-to-zero-with-keda/README.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+Once you generate a load ( even a single request ) the non-zero request rate targeting the worker Function will trigger scaling up of the worker Function's runtime Pods.
+
+ Call the http proxy Function once 
+
+ ```bash
+ curl -H "Content-Type: application/cloudevents+json" -X POST -d '{"foo":"bar"}' https://incoming.{your_cluster_domain}
+ ```
+
+The message will be pushed to the kyma eventing.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:67]:
```suggestion
The message is pushed to the Kyma Eventing.
```

File: scale-to-zero-with-keda/README.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+Ofcourse it takes time to scale up a Function from zero. But no message will be lost as eventing will retry delivering message to subscriber until it is eventually consumed bu a running worker Pod.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:68]:
```suggestion
It takes time to scale up a Function from zero. But no message is lost as Eventing retries delivery of the message to the subscriber until a running worker Pod eventually consumes it.
```

File: scale-to-zero-with-keda/README.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+Observe worker Function scaling up from 0. You will notice by watching Function Pods or HPA.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:70]:
```suggestion
Observe worker Function scaling up from zero. You can notice it by watching Function Pods or HPA.
```

File: scale-to-zero-with-keda/README.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+scalable-worker-fn-2n269-6f6d5f675-t6nwr   0/2     Running           0          7s
+```
+```bash
+kubectl get hpa -w                                                        
+NAME                               REFERENCE                     TARGETS             MINPODS   MAXPODS   REPLICAS   AGE
+keda-hpa-worker-fn-scaled-object   Function/scalable-worker-fn   <unknown>/2 (avg)   1         5         0          27h
+keda-hpa-worker-fn-scaled-object   Function/scalable-worker-fn   0/2 (avg)           1         5         1          27h
+```
+
+Observe that the payload was eventually processed
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:90]:
```suggestion
Observe that the payload was eventually processed.
```

File: scale-to-zero-with-keda/README.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+See in the worker's Function logs that `Processing ... {"foo":"bar"}` eventually appears:
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:92]:
```suggestion
Check the worker Function logs. `Processing ... {"foo":"bar"}` eventually appears:
```

File: scale-to-zero-with-keda/README.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+> nodejs16-runtime@0.1.0 start
+> node server.js
+
+user code loaded in 0sec 0.783514ms
+Processing ...
+{"foo":"bar"}
+
+ ```
+ 
+ If the trafic stops, the worker Function will be scaled down back to zero replicas (after a configurable cooldown period)
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:105]:
```suggestion
 If the traffic stops, the worker Function is scaled down back to zero replicas (after a configurable cooldown period)
```

File: scale-to-zero-with-keda/README.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+ 
+ If you generate a much higher load (for example > 2 req/sec - as configured in the threshold value of the scaledObject) you will observe scaling up to more replicas. One replica should be added for each additional 2req/sec measured.
```
Comment (grego952) [inline scale-to-zero-with-keda/README.md:107]:
```suggestion
 If you generate a much higher load, for example,> 2 req/sec - as configured in the threshold value of the scaledObject, you will observe scaling up to more replicas. One replica should be added for each additional 2req/sec measured. 
```

File: scale-to-zero-with-keda/README.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+Run the following against your Kyma runtime to install KEDA:
+
+```bash
+kubectl apply -f https://github.com/kedacore/keda/releases/download/v2.8.0/keda-2.8.0.yaml
+```
+
+Make sure istio sidecar injection is enabled in the target Namesapce:
+
+```bash
+kubectl label ns default istio-injection=enabled
```
Comment (dbadura) [inline scale-to-zero-with-keda/README.md:34]:
```suggestion
kubectl label namespace default istio-injection=enabled
```

**Review by grego952 — DISMISSED** [review]

**Review by grego952 — APPROVED** [review]

**Pull request merged**